### ISSUE TITLE
compaction: contended per-key atomic store limits scalability of concurrent compactions

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/pebble#1808]
**TL;DR: an unnecessary [per-key atomic store](https://github.com/cockroachdb/pebble/blob/f30672e7240bf0259d9eacf62111d6df81527d6d/compaction.go#L2370) on a global memory address during compactions slows down large index backfills by about 10%.**

During an index creation on a 6TB table in a cluster with 15 nodes (48 vCPU, single NUMA node per instance), I noticed a strange pattern in CPU profiles across the cluster. Each node was pushing about 300 MB/s of compactions and consuming about 35% of its CPU. This was all after having increased `COCKROACH_ROCKSDB_CONCURRENCY` to 16 to better utilize the hardware with up to 16 concurrent compactions, instead of the default of 4.

Each time I captured a CPU profile ([compaction_profile.pb.gz](https://github.com/cockroachdb/pebble/files/9079875/compaction_profile.pb.gz)) to determine where CPU time was being used, I saw the following in the `peek` output:
```
Showing nodes accounting for 98.85s, 100% of 98.85s total
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context
----------------------------------------------------------+-------------
                                            13.42s 95.18% |   github.com/cockroachdb/pebble.(*DB).compact1 github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1888
                                             0.68s  4.82% |   github.com/cockroachdb/pebble.(*DB).flush1 github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1441
    14.10s 14.26% 14.26%     14.10s 14.26%                | github.com/cockroachdb/pebble.(*DB).runCompaction github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:2351
----------------------------------------------------------+-------------
```

Notice that the single largest consumer of CPU (14.26% of the total profile) was `pebble/compaction.go:2351`. This is:
https://github.com/cockroachdb/pebble/blob/f30672e7240bf0259d9eacf62111d6df81527d6d/compaction.go#L2351

This made little sense — why was an interface comparison so expensive? Interestingly, we've seen this before in https://github.com/cockroachdb/pebble/issues/1030, and landed both https://github.com/cockroachdb/pebble/pull/1059 and https://github.com/cockroachdb/pebble/pull/1160 in response. Why was is still so expensive?

Looking at the `disasm` profile helped answer that. The instructions that correspond to this comparison were cold.
```
         .          .    11c21cd: movq	3808(%rsp), %rdx                   ;github.com/cockroachdb/pebble.(*DB).runCompaction compaction.go:2351
         .          .    11c21d5: testq	%rdx, %rdx
         .          .    11c21d8: je	0x11c21fc <github.com/cockroachdb/pebble.(*DB).runCompaction+0x195c>
         .          .    11c21da: leaq	55784575(%rip), %rbx    # 0x46f5660 <type.*+0x372d80>
         .          .    11c21e1: cmpq	%rbx, 8(%rdx)
         .          .    11c21e5: jne	0x11c21fc <github.com/cockroachdb/pebble.(*DB).runCompaction+0x195c>
         .          .    11c21e7: movq	912(%rsp), %rbx
         .          .    11c21ef: cmpq	%rbx, 132892394(%rip)   # 0x907e8e0 <github.com/cockroachdb/pebble.nilPacer>
         .          .    11c21f6: sete	%sil
         .          .    11c21fa: jmp	0x11c21fe <github.com/cockroachdb/pebble.(*DB).runCompaction+0x195e>
         .          .    11c21fc: xorl	%esi, %esi
         .          .    11c21fe: movb	%sil, 186(%rsp)
```

However, later down in the function, we see:
```
         .      4.02s    11c2540: callq 0x11b3c60 <github.com/cockroachdb/pebble.(*splitterGroup).shouldSplitBefore>
      10ms       10ms    11c2545: cmpq  $1, %rax
         .          .    11c2549: je    0x11c23c9 <github.com/cockroachdb/pebble.(*DB).runCompaction+0x1b29>
         .          .    11c254f: movq  3800(%rsp), %r11                   ;github.com/cockroachdb/pebble.(*DB).runCompaction compaction.go:2370
         .          .    11c2557: movq  168(%r11), %r12
      20ms       20ms    11c255e: movq  152(%r11), %r13
         .          .    11c2565: xchgq %r13, (%r12)
    14.08s     14.08s    11c2569: movzbl        186(%rsp), %r12d                 ;github.com/cockroachdb/pebble.(*DB).runCompaction compaction.go:2351
      20ms       20ms    11c2572: testb %r12b, %r12b
                                                                          ;github.com/cockroachdb/pebble.(*DB).runCompaction compaction.go:2371
         .          .    11c2575: jne   0x11c25ae <github.com/cockroachdb/pebble.(*DB).runCompaction+0x1d0e>
```

Notice that the hot `movzbl` instruction is attributed to line 2351, even though it comes after line 2370. So the use of the hoisted `isNilPacer` variable is expensive and that cost is attributed back to line 2351:
https://github.com/cockroachdb/pebble/blob/f30672e7240bf0259d9eacf62111d6df81527d6d/compaction.go#L2371

On a hunch, I guessed that the time was being either misattributed, either directly due to a bug with profiling or because something else was causing the memory load to be more expensive. I'm still not sure whether either of these guesses were correct, but this led to me questioning the previous line:
https://github.com/cockroachdb/pebble/blob/f30672e7240bf0259d9eacf62111d6df81527d6d/compaction.go#L2370

This is an atomic store for each key in the compaction. Why is it atomic? Is `c.atomicBytesIterated` not local to the current compaction? A bit of digging reveals that the atomic store was added in https://github.com/cockroachdb/pebble/pull/201 and that `c.atomicBytesIterated` is aliased to either `DB.bytesFlushed` or `DB.bytesCompacted` at any point in time. In other words, this is an atomic store on a memory address that is shared between all compactions. Furthermore, it's an atomic memory store that is performed for each key in each compaction. That sounds expensive.

This raises two questions:
1. is this needed?
2. is this correct?

I believe the answer to both questions is "no". Tracking these values is only needed for the `compactionPacer`, which is disabled, so this is not needed. The use of a global memory address also assumes that only a single compaction is running at a time. This is why writes to the values are stores instead of something commutative like an atomic increment. If two compactions run at the same time, they will clobber each other's values, so this can't possibly be working today.

To confirm this was the real source of CPU usage, especially on large machines with many concurrent compactions and where contention on the memory addresses becomes a large issue, I patched pebble to remove these atomic stores (https://github.com/nvanbenschoten/pebble/commit/641915ef6e0e540234f112d94f633716e529789a). Using a binary compiled without this patch and one compiled with this patch, I ran the following test:
```
# create cluster and stage
version=before
cluster=${USER}-import-${version}
roachprod create ${cluster} --nodes=3 --gce-machine-type=n2-standard-48 --gce-zones=us-central1-a --local-ssd=false --gce-pd-volume-size=5000 --gce-min-cpu-platform='Intel Ice Lake'
roachprod put ${cluster} cockroach-${version} cockroach
roachprod start ${cluster} --env=COCKROACH_ROCKSDB_CONCURRENCY=16 --env=COCKROACH_ENABLE_RPC_COMPRESSION=false

# import and backfill
roachprod run ${cluster}:1
tmux
./cockroach sql --insecure -e "CREATE TABLE defaultdb.public.trade (t_id INT8 NOT NULL PRIMARY KEY, t_dts TIMESTAMP NOT NULL, t_st_id VARCHAR(4) NOT NULL, t_tt_id VARCHAR(3) NOT NULL, t_is_cash BOOL NOT NULL, t_s_symb VARCHAR(15) NOT NULL, t_qty INT4 NOT NULL CHECK (t_qty > 0), t_bid_price DECIMAL(8,2) NOT NULL CHECK (t_bid_price > 0), t_ca_id INT8 NOT NULL, t_exec_name VARCHAR(49) NOT NULL, t_trade_price DECIMAL(8,2), t_chrg DECIMAL(10,2) NOT NULL CHECK (t_chrg >= 0), t_comm DECIMAL(10,2) NOT NULL CHECK (t_comm >= 0), t_tax DECIMAL(10,2) NOT NULL CHECK (t_tax  >= 0), t_lifo BOOL NOT NULL, FAMILY (t_id), FAMILY (t_comm, t_dts, t_st_id, t_trade_price, t_exec_name, t_tt_id, t_is_cash, t_s_symb, t_qty, t_bid_price, t_ca_id, t_chrg, t_lifo), FAMILY (t_tax))"
./cockroach sql --insecure -e "IMPORT INTO defaultdb.public.trade CSV DATA ('gs://cockroach-fixtures/tpce-csv/customers=500000/*/Trade.txt?AUTH=implicit') WITH delimiter = '|'";
time ./cockroach sql --insecure -e "CREATE INDEX ON defaultdb.public.trade (t_s_symb, t_dts ASC) STORING (t_ca_id, t_exec_name, t_is_cash, t_trade_price, t_qty, t_tt_id)";
```

This test confirmed the suspicion. Without the patch, CPU profiles ([before_profile.pb.gz](https://github.com/cockroachdb/pebble/files/9079894/before_profile.pb.gz)) still show the original `isNilPacer := pacer == nilPacer` line to be the single largest source of CPU time. With the patch, that line is nowhere to be found in the CPU profiles ([after_profile.pb.gz](https://github.com/cockroachdb/pebble/files/9079895/after_profile.pb.gz)).

Furthermore, the index backfill was measurably faster with the patch. Without the patch, the index creation took **09:10:53**. After the patch, the index creation took **08:21:37**. The patch reduced the duration of the entire index creation by **8.94%**.

Of course, the atomic store is more contended with 16 concurrent compactions than it would have been with 4. Still, we have plenty of evidence (https://github.com/cockroachdb/pebble/commit/91c52f6052eed0573fab4aad266a1d18f2510ad5) that this was a problem even with less compaction concurrency.

We should determine what our plans are with the pacers. If they're needed, let's fix them and avoid this faulty global accounting of bytes flushed and compacted. If they're not needed, should we rip them out?

**Comment by jbowens** [comment]
Thanks @nvanbenschoten for digging into this. Great find. I agree this is not needed, and is unlikely to be needed soon [ever?].

My reading of the code is that `DB.bytesCompacted` was unused, even by flush/compaction pacing. It served purely as a non-nil value to provide during compactions, needlessly adding contention. The `DB.bytesFlushed` value _was_ used by pacing. Because flushes are single-threaded, the atomic store was "correct" but unused with pacing disabled.

I think we should strip out the pacing and rely on our resource-aware pacing (#1329). If we ever change our minds and decide there's something to be gained by reviving this particular implementation of pacing, we can pay the cost of rebasing/reintroducing the code then.

I put up #1812 with two commits: the first simply removes the atomic store, like your commit. The second removes compaction and flush pacing altogether.
